plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/IngestActionFilter.java

```diff
@@ -137,4 +151,91 @@ public void failed(Exception e) {
     public int order() {
         return Integer.MAX_VALUE;
     }
+
+    final static class BulkRequestSource implements Iterator<ActionRequest> {
+
+        final BulkRequest bulkRequest;
+        final ActionListener<BulkResponse> actionListener;
+
+        final Set<Integer> failedSlots;
+        final List<BulkItemResponse> itemResponses;
+
+        int currentSlot = -1;
+
+        BulkRequestSource(BulkRequest bulkRequest, ActionListener<BulkResponse> actionListener) {
+            this.bulkRequest = bulkRequest;
+            this.actionListener = actionListener;
+            this.failedSlots = new HashSet<>();
+            this.itemResponses = new ArrayList<>(bulkRequest.requests().size());
+        }
+
+        @Override
+        public ActionRequest next() {
+            return bulkRequest.requests().get(++currentSlot);
+        }
+
+        @Override
+        public boolean hasNext() {
+            return (currentSlot + 1) < bulkRequest.requests().size();
+        }
+
+        Tuple<BulkRequest, ActionListener<BulkResponse>> bulkRequestAndActionListener() {
```

Comment:
The way I see this is because `BulkRequestSource` already holds the bulk request and failed slots, it makes sense to me to to add any helper methods that work with these things there instead of placing it somewhere else.